workflow/controller/controller.go

```diff
@@ -522,6 +515,58 @@ func (wfc *WorkflowController) runPodCleanup(ctx context.Context) {
 	}
 }
 
+func (wfc *WorkflowController) getPodCleanupPatch(pod *apiv1.Pod, labelPodCompleted bool) ([]byte, error) {
+	un := unstructured.Unstructured{}
+	if labelPodCompleted {
+		un.SetLabels(map[string]string{common.LabelKeyCompleted: "true"})
+	}
+
+	finalizerEnabled := os.Getenv("ARGO_POD_STATUS_CAPTURE_FINALIZER") == "true"
+	if finalizerEnabled && slice.ContainsString(pod.Finalizers, common.FinalizerPodStatus) {
+		un.SetFinalizers(slice.RemoveString(pod.Finalizers, common.FinalizerPodStatus))
```

Comment:
this also does still loop over finalizers twice like I mentioned [before](https://github.com/argoproj/argo-workflows/pull/12862#discussion_r1555119124).